qp_klp/klp.py

```diff
@@ -40,66 +45,180 @@ def sequence_processing_pipeline(qclient, job_id, parameters, out_dir):
+            qurl = f'/api/v1/study/{qiita_id}/samples'
+
+            if tids:
+                # strip any leading zeroes from the sample-ids. Note that
+                # if a sample-id has more than one leading zero, all of
+                # them will be removed.
+                my_samples = {x.lstrip('0') for x in samples}
+                # once any leading zeros have been removed, recalculate
+                # sample_name_diff before continuing processing.
+                sample_name_diff = my_samples - qsam
+
+                # before we report as an error, check tube_id.
+                if pipeline_type == 'metagenomic':
+                    tids = qclient.get(f'{qurl}/categories=tube_id')[
+                        'samples']
+
+                # generate a map of sample_names to tube_ids for
+                # GenPrepFileJob.
+                sn_tid_map_by_project[project_name] = {
+                    y[0]: x.replace(f'{qiita_id}.', '') for x, y in
+                    tids.items()}
+                tids = set(sn_tid_map_by_project[project_name].keys())
+                tube_id_diff = my_samples - tids
+                if not tube_id_diff:
+                    continue
+                len_tube_id_overlap = len(tube_id_diff)
+                tidsx = ', '.join(
+                    tids if len(tids) < 6 else rsampl(tids, k=5))
+
+                error_tube_id = (
+                    f'tube_id in Qiita but {len_tube_id_overlap} missing '
+                    f'samples. Some samples from tube_id: {tidsx}.')
+
+            len_overlap = len(sample_name_diff)
+            # selecting at random k=5 samples to minimize space in display
+            samx = ', '.join(qsam if len(qsam) < 6 else rsampl(qsam, k=5))
+
+            # selecting the up to 4 first samples to minimize space in
+            # display
+            missing = ', '.join(sorted(sample_name_diff)[:4])
+            errors.append(
+                f'{project_name} has {len_overlap} missing samples (i.e. '
+                f'{missing}). Some samples from Qiita: {samx}. '
+                f'{error_tube_id}')
+
+        if errors:
+            raise PipelineError('\n'.join(errors))
+
+        # find the uploads directory all trimmed files will need to be
+        # moved to and generate a map.
+        special_map = generate_special_map(
+            qclient.get("/qiita_db/artifacts/types/"),
+            pipeline.get_project_info())
+
+        if pipeline.pipeline_type == 'metagenomic':
+            step = MetagenomicStep(pipeline, job_id, sn_tid_map_by_project,
+                                   status_line)
+        else:
+            # pipeline.pipeline_type == 'amplicon':
+            step = AmpliconStep(pipeline, job_id, sn_tid_map_by_project,
+                                status_line)
+
+        step.convert_bcl_to_fastq()
+
+        step.quality_control()
+
+        step.generate_reports()
+
+        step.generate_prep_file()
+
+        from_qiita = {}
```

Comment:
Looking at this and the code in Step is clear that this is not a step but a controller, similar to what you used to have of Pipeline; is that right? If yes, why the rename?

Reply:
Not like Pipeline. 

> Thank you for the comments before. Now, a general question, if you are "extracting" the general functions from Amplicon/Metagenomic Step and adding them to the base class Step, does it really make sense to have a klp_util.py file? I mean basically you have to access method X in Amplicon/Metagenomic Step, that will call X method in Step and then do X in klp_util.py; no? Would it be better to remove one step and just go either Amplicon/Metagenomic Step -> klp_util.py or Amplicon/Metagenomic Step -> Step and remove klp_util.py?

We're definitely on the same page. The idea was to incorporate klp_util.py into the base Step class. I left them as is for the time being, just because so many other things had changed already.